datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -1503,12 +1579,39 @@ fn get_buffered_columns(
     buffered_batch_idx: usize,
     buffered_indices: &UInt64Array,
 ) -> Result<Vec<ArrayRef>, ArrowError> {
-    buffered_data.batches[buffered_batch_idx]
-        .batch
-        .columns()
-        .iter()
-        .map(|column| take(column, &buffered_indices, None))
-        .collect::<Result<Vec<_>, ArrowError>>()
+    get_buffered_columns_from_batch(
+        &buffered_data.batches[buffered_batch_idx],
+        buffered_indices,
+    )
+}
+
+#[inline(always)]
+fn get_buffered_columns_from_batch(
+    buffered_batch: &BufferedBatch,
+    buffered_indices: &UInt64Array,
+) -> Result<Vec<ArrayRef>, ArrowError> {
+    if let Some(spill_file) = &buffered_batch.spill_file {
+        // if spilled read from disk in smaller sub batches
+        let mut buffered_cols: Vec<ArrayRef> = Vec::with_capacity(buffered_indices.len());
+
+        let file = BufReader::new(File::open(spill_file.path())?);
+        let reader = FileReader::try_new(file, None)?;
+
+        for batch in reader {
+            batch?.columns().iter().for_each(|column| {
+                buffered_cols.extend(take(column, &buffered_indices, None))
```

Comment:
Is this right to take the indexes on each buffered batch? It seems like the indices need to account for the fact that multiple batches got read in from the spill file

Or put another way, does `buffered_indices` refer to the overall logical indices within the `BufferedBatch` or is it within each potential `RecordBatch` read from the spill file?

Reply:
On this I feel I need to think more. you are right if we save data on smaller batches the indices we looking up can return wrong results across multiple  batches

Reply:
It should be okay, as buffered indices are created by the same algorithm of splitting incoming buffer batch by the batch size (https://github.com/apache/datafusion/blob/main/datafusion/physical-plan/src/joins/sort_merge_join.rs#L1065) the similar as we split buffered batch into subbatches during spilling by the batch size. When it comes to read phase the data should be consistent.

But also to feel more confident  I have added for spill tests more assertion so now it also checks result between spilled and non spilled SMJ runs